Add a sanity test before formatting Q# code

yjt98765 opened the pull request:
This pull request implements issue #867, i.e., it unparses the syntax tree and compares the result to the original code before applying format rules.

## Modification
To use the `parse` variable, I have to change the accessibility of `Document`. In addition, the F# compiler also requires me to change the accessibility of `Terminal`, which I do not quite understand.

When the formatter fails the sanity test, it reports an error message. The message is composed based on the [README](https://github.com/microsoft/qsharp-compiler/blob/main/README.md#reporting-security-issues) file.

## Testing
I have to change the `Formats standard input` test after implementing the above feature. Because I am working on Windows, the unparsing result ends with `\r\n`, which is different from the '\n' ending in the test case. This difference makes the test fail.

The above incident inspires me to add a new test named `Shows unparsed tree not equal to source error` to [Tests.qs](https://github.com/microsoft/qsharp-compiler/blob/main/src/QsFmt/App.Tests/Tests.fs). In the Q# code attached to this test, I inserted both `\r\n` and `\n`. It ensures that the unparsed result in all environments will be different from the source. In this way, it can trigger the newly added error message.

I am not sure whether not being able to handle mixed line endings is a bug. In most cases, it will not exist in users' code. However, I tried similar things to test F#'s formatter. It can still format the file without reporting errors. Maybe we should omit the line-ending differences in the sanity test?

samarsha commented on src/QsFmt/App/Program.fs:
```diff
@@ -29,13 +29,24 @@ let main args =
+        match Formatter.parse source with
+        | Ok document ->
+            // Test whether there is data loss during parsing and unparsing
+            if Formatter.unparse document = source then
```
It might be simpler to move this check inside the `Formatter.format` function. I think that you wouldn't need to expose new functions or types in that case.

yjt98765 replied:
Agree. I have rolled back most changes in my last commit, i.e., I have added back `internal` labels and removed new exposing functions. Now, the check is within `Formatter.format` function.

samarsha reviewed (commented):
> I have to change the `Formats standard input` test after implementing the above feature. Because I am working on Windows, the unparsing result ends with `\r\n`, which is different from the '\n' ending in the test case. This difference makes the test fail.

Oh, I see. I guess that is because the input string uses `\n` but the expected output string uses an actual newline in the source code. I think your change to use the same type of newline in both strings makes sense.

> I am not sure whether not being able to handle mixed line endings is a bug. In most cases, it will not exist in users' code. However, I tried similar things to test F#'s formatter. It can still format the file without reporting errors. Maybe we should omit the line-ending differences in the sanity test?

Hmm, I think that if there are mixed newlines in the original source code, then parsing and unparsing should preserve the original newlines. So we should still be able to compare the strings for an exact match. Is that not what is happening?

yjt98765 commented:
> > I am not sure whether not being able to handle mixed line endings is a bug. In most cases, it will not exist in users' code. However, I tried similar things to test F#'s formatter. It can still format the file without reporting errors. Maybe we should omit the line-ending differences in the sanity test?
> 
> Hmm, I think that if there are mixed newlines in the original source code, then parsing and unparsing should preserve the original newlines. So we should still be able to compare the strings for an exact match. Is that not what is happening?

No. In the [original code](https://github.com/microsoft/qsharp-compiler/pull/1084/files#diff-d25afa6d6770df0d965521635ea3fbb306ae5c6faaedd4d359ba950ee6f7de8fL16), all newlines are printed as `Environment.NewLine` instead of original ones. However, I have changed it so that the original newlines can be preserved now.

The "downside" of this change is that the [test case](https://github.com/microsoft/qsharp-compiler/pull/1084/commits/75c0b780fdd904ac23038699bb328d4793b794dc#diff-0b2979608a6ed36fc432c25f756a899b5fe480ac324b322ae79e1f21f463f7e8L133-L145) I added last time is invalid now. I cannot think of a code example to trigger the bug [reporting part](https://github.com/microsoft/qsharp-compiler/pull/1084/commits/75c0b780fdd904ac23038699bb328d4793b794dc#diff-87778067becd07ddecda46d7e5a1a3fb1ef7bb32095b43f07f514e57d4e4a359L41-L45). Even if we find one, it means there is a bug in our code, and thus it should be resolved quickly. How can we test the sanity test? Currently, all the test cases pass.

samarsha commented on src/QsFmt/App/Program.fs:
```diff
@@ -46,3 +47,6 @@ let main args =
         eprintfn "%s" ex.Message
         4
+    | Failure (msg) ->
+        eprintfn "%s" msg
```
This hides the stack trace which may be useful for debugging (and to include in a bug report). I would be fine with leaving this as an unhandled exception, since the default behavior is to print the message, stack trace, and (on Windows) show a message box saying the program stopped working.

samarsha commented on src/QsFmt/Formatter/Formatter.fs:
```diff
@@ -33,14 +33,24 @@ let parse (source: string) =
+        else
+            failwith (
+                "The formater does not work properly. "
+                + "Please let us know by filing a new issue in https://github.com/microsoft/qsharp-compiler/issues/new/choose."
```
It would be nice if the error message was a little more specific - for example it could say that the formatter's syntax tree is not consistent with the input source code. Also, including the unparsed document in the message could help with debugging.

samarsha commented on src/QsFmt/Formatter/SyntaxTree/Node.fsi:
```diff
@@ -37,6 +37,9 @@ module internal Trivia =
     /// </summary>
     val newLine: Trivia
```
It might be good to update the documentation for `newLine` to clarify that it's a new line node containing the default new line character for the current platform.

samarsha reviewed (commented):
> No. In the [original code](https://github.com/microsoft/qsharp-compiler/pull/1084/files#diff-d25afa6d6770df0d965521635ea3fbb306ae5c6faaedd4d359ba950ee6f7de8fL16), all newlines are printed as `Environment.NewLine` instead of original ones. However, I have changed it so that the original newlines can be preserved now. See the newly edited [test case](https://github.com/microsoft/qsharp-compiler/pull/1084/files#diff-0b2979608a6ed36fc432c25f756a899b5fe480ac324b322ae79e1f21f463f7e8L93).

You're right, I forgot about that, thanks for fixing it. :)

> I cannot think of a code example to trigger the bug [reporting part](https://github.com/microsoft/qsharp-compiler/pull/1084/commits/75c0b780fdd904ac23038699bb328d4793b794dc#diff-87778067becd07ddecda46d7e5a1a3fb1ef7bb32095b43f07f514e57d4e4a359L41-L45). Even if we find one, it means there is a bug in our code, and thus it should be resolved quickly. How can we test the sanity test? Currently, all the test cases pass.

Yes, I think I'm OK with not having a test for this error message. By design there should be no input that triggers the error, and it doesn't seem like a good idea to add a way to trigger the error manually.

yjt98765 commented on src/QsFmt/App.Tests/Tests.fs:
```diff
@@ -90,13 +90,11 @@ let ``Formats file`` path output =
 
 [<Theory>]
-[<InlineData("namespace Foo { function Bar() : Int { return 0; } }\n",
+[<InlineData("namespace Foo { function Bar() : Int { return 0; }\n }\r\n",
```
Sure. I have added the [new test](https://github.com/microsoft/qsharp-compiler/pull/1084/files#diff-27cec60c6dde2dea728bb9a23831ee402cd9639782130ba7112f37acc56bf1e4R46-R50) in FixedPoints.fs.

ScottCarda-MS commented on src/QsFmt/Formatter/Printer.fs:
```diff
@@ -5,15 +5,14 @@ module Microsoft.Quantum.QsFmt.Formatter.Printer
     function
     | Whitespace ws -> ws
-    | NewLine -> Environment.NewLine
+    | NewLine nl -> nl
```
I think should keep this as Environment.NewLine to keep this flexible with different local setups. My local setup, for example is a mix of different new line characters, and having this be a specific new line, one way or the other, breaks most of my tests. My recommendation is to leave this as Environment.NewLine and instead use a function like the following on inputs before string-comparison:
```
/// <summary>
/// Replaces New Line characters in source string for Environment.NewLine characters.
/// </summary>
let private standardizeNewLines (source: string) =
    source.Replace("\r", "").Replace("\n", Environment.NewLine)
```

samarsha reviewed (approved):
Looks great, thanks!

samarsha commented:
/azp run
